File: hololive/hololive-rs/crates/shared/core/src/model/command.rs
use std::collections::HashMap;

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum CommandType {
    #[serde(rename = "live")]
    Live,
    #[serde(rename = "upcoming")]
    Upcoming,
    #[serde(rename = "schedule")]
    Schedule,
    #[serde(rename = "help")]
    Help,
    #[serde(rename = "alarm_add")]
    AlarmAdd,
    #[serde(rename = "alarm_remove")]
    AlarmRemove,
    #[serde(rename = "alarm_list")]
    AlarmList,
    #[serde(rename = "alarm_clear")]
    AlarmClear,
    #[serde(rename = "alarm_invalid")]
    AlarmInvalid,
    #[serde(rename = "member_info")]
    MemberInfo,
    #[serde(rename = "stats")]
    Stats,
    #[serde(rename = "subscriber")]
    Subscriber,
    #[serde(rename = "member_news")]
    MemberNews,
    #[serde(rename = "news_subscription")]
    MemberNewsSubscription,
    #[serde(rename = "major_event")]
    MajorEvent,
    #[serde(rename = "unknown")]
    Unknown,
}

impl CommandType {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Live => "live",
            Self::Upcoming => "upcoming",
            Self::Schedule => "schedule",
            Self::Help => "help",
            Self::AlarmAdd => "alarm_add",
            Self::AlarmRemove => "alarm_remove",
            Self::AlarmList => "alarm_list",
            Self::AlarmClear => "alarm_clear",
            Self::AlarmInvalid => "alarm_invalid",
            Self::MemberInfo => "member_info",
            Self::Stats => "stats",
            Self::Subscriber => "subscriber",
            Self::MemberNews => "member_news",
            Self::MemberNewsSubscription => "news_subscription",
            Self::MajorEvent => "major_event",
            Self::Unknown => "unknown",
        }
    }

    pub fn is_valid(self) -> bool {
        true
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParseResult {
    pub command: CommandType,
    pub params: HashMap<String, serde_json::Value>,
    pub confidence: f64,
    pub reasoning: String,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ParseResults {
    pub single: Option<ParseResult>,
    pub multiple: Vec<ParseResult>,
}

impl ParseResults {
    pub fn is_single(&self) -> bool {
        self.single.is_some()
    }

    pub fn is_multiple(&self) -> bool {
        !self.multiple.is_empty()
    }

    pub fn commands(&self) -> Vec<&ParseResult> {
        if let Some(single) = &self.single {
            return vec![single];
        }

        self.multiple.iter().collect()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CommandContext {
    pub room: String,
    pub room_name: String,
    pub user_id: String,
    pub user_name: String,
    pub is_group_chat: bool,
    pub message: String,
    pub timestamp: DateTime<Utc>,
}

File: hololive/hololive-shared/pkg/domain/command.go
// Copyright (c) 2025 Kapu
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package domain

type CommandType string

const (
	CommandLive                   CommandType = "live"
	CommandUpcoming               CommandType = "upcoming"
	CommandSchedule               CommandType = "schedule"
	CommandHelp                   CommandType = "help"
	CommandAlarmAdd               CommandType = "alarm_add"
	CommandAlarmRemove            CommandType = "alarm_remove"
	CommandAlarmList              CommandType = "alarm_list"
	CommandAlarmClear             CommandType = "alarm_clear"
	CommandAlarmInvalid           CommandType = "alarm_invalid"
	CommandMemberInfo             CommandType = "member_info"
	CommandStats                  CommandType = "stats"
	CommandSubscriber             CommandType = "subscriber"
	CommandMemberNews             CommandType = "member_news"
	CommandMemberNewsSubscription CommandType = "news_subscription"
	CommandMajorEvent             CommandType = "major_event"
	CommandCalendar               CommandType = "calendar"
	CommandBroadcastHistory       CommandType = "broadcast_history"
	CommandBroadcastThumbnail     CommandType = "broadcast_thumbnail"
	CommandUnknown                CommandType = "unknown"
)

func (c CommandType) String() string {
	return string(c)
}

func (c CommandType) IsValid() bool {
	switch c {
	case CommandLive, CommandUpcoming, CommandSchedule, CommandHelp,
		CommandAlarmAdd, CommandAlarmRemove, CommandAlarmList, CommandAlarmClear, CommandAlarmInvalid,
		CommandMemberInfo, CommandStats, CommandSubscriber,
		CommandMemberNews, CommandMemberNewsSubscription,
		CommandMajorEvent,
		CommandCalendar,
		CommandBroadcastHistory, CommandBroadcastThumbnail,
		CommandUnknown:
		return true
	default:
		return false
	}
}

type ParseResult struct {
	Command    CommandType    `json:"command"`
	Params     map[string]any `json:"params"`
	Confidence float64        `json:"confidence"`
	Reasoning  string         `json:"reasoning"`
}

type ParseResults struct {
	Single   *ParseResult
	Multiple []*ParseResult
}

type ChannelSelection struct {
	SelectedIndex int     `json:"selectedIndex"`
	Confidence    float64 `json:"confidence"`
	Reasoning     string  `json:"reasoning"`
}

func (pr *ParseResults) IsSingle() bool {
	return pr.Single != nil
}

func (pr *ParseResults) IsMultiple() bool {
	return len(pr.Multiple) > 0
}

func (pr *ParseResults) GetCommands() []*ParseResult {
	if pr.IsSingle() {
		return []*ParseResult{pr.Single}
	}
	return pr.Multiple
}
